lapce-ui/src/editor/view.rs

```diff
@@ -259,19 +259,51 @@ impl LapceEditorView {
         }
     }
 
-    pub fn ensure_cursor_center(
+    pub fn ensure_cursor_position(
         &mut self,
         ctx: &mut EventCtx,
         data: &LapceEditorBufferData,
         panels: &im::HashMap<PanelPosition, Arc<PanelData>>,
+        position: &EnsureVisiblePosition,
         env: &Env,
     ) {
-        let center = Self::cursor_region(data, ctx.text()).center();
-
-        let rect = Rect::ZERO.with_origin(center).inflate(
-            (data.editor.size.borrow().width / 2.0).ceil(),
-            (data.editor.size.borrow().height / 2.0).ceil(),
-        );
+        // This is where the cursor currently is.
+        let cursor = Self::cursor_region(data, ctx.text()).center();
+        // FIXME: need to see how LapceEditorBufferData::scroll() handles the
+        // margin
+        //
+        // TODO: scroll margin (in number of lines) should be configurable
+        let margin = data.config.editor.line_height * 2;
+
+        // The origin of a rect is its top-left corner.  Inflating a point
+        // creates a rect that centers at the point.
+        let width = (data.editor.size.borrow().width / 2.0).ceil();
+        let height = (data.editor.size.borrow().height / 2.0).ceil();
+
+        // Find where the center will be.
+        let center = match position {
+            EnsureVisiblePosition::CenterOfWindow => {
+                // Cursor line will be at the center of the view.
+                cursor
+            },
+            EnsureVisiblePosition::TopOfWindow => {
+                // Cursor line will be at the top edge, thus center of the view
+                // is half height down.
+                let h = height as usize;
+                let h = if h > margin { h - margin } else { margin };
+                Point::new(cursor.x, cursor.y + h as f64)
+            },
+            EnsureVisiblePosition::BottomOfWindow => {
+                // Cursor line will be at the bottom edge, thus center of the
+                // view is half height up.
+                let h = height as usize;
+                let h = if h > margin { h - margin } else { margin };
+                let y = cursor.y as usize;
+                let y = if h > y { y } else { y - h };
```

Comment:
This pattern is repeated often enough to deserve its own function. This is basically `saturating_sub`, right? Also I think this code is complicated enough to warrant some tests.

Reply:
> This pattern is repeated often enough to deserve its own function. ~This is basically `saturating_sub`, right?~ No it's not. 

Indeed the calculation was incorrect. Fixed in the later commits.

> Anyway, I also think this code is complicated enough to warrant some tests.

I assume that you meant unit tests? I was trying to add some but could not figure out how to verify it without repeating the same calculation in the tests. I need to think about it.